### ISSUE TITLE
Investigate memory consumption when compressing large files

### ISSUE THREAD

**Issue by mhx** [mhx/dwarfs#35]
> I have to admit I've been doing most of my compression tasks on machines with 64GB of memory, so optimizing for low memory consumption hasn't really been a priority yet. There are some knobs you might be able to turn, though. I'm not sure large files per se are an issue, but a large number of files definitely is. You might be able to tweak `--memory-limit` a bit, which determines how many uncompressed blocks can be queued. If you lower this limit, the compressor pool may run out of blocks more quickly, resulting in overall slower compression. Reducing the number of workers (`-N`) might also help a bit.

A small update on this (apologies that this is on an unrelated issue). Did some experimentation and found that lowering memory limit and workers works in some instances but not in others. Large files seems to be the biggest hold up, in particular an instance when I tried to put a 3.1 GB file which seemingly had no way of compressing via dwarfs with my 16 GB of memory (even with very low options like `-L1m -N1`).

What I did find instead was that using the `-l0` option and then recompressing the image works in these cases without issue. Creating the initial image with `-S24` results in very well recompressed files in these instances, the 3.1 GB file compressing down to 2.3 GB whereas the default block size for `-l0` resulted in a 2.6 GB file (which is approx what `mksquashfs -comp zstd -b 1M -Xcompression-level 22` also gave me).

_Originally posted by @Phantop in https://github.com/mhx/dwarfs/issues/33#issuecomment-778908533_

**Comment by mhx** [comment]
@Phantop, just to make sure I understand this correctly: you're trying to build DwarFS image from a tree where single files have a size of 3.1GB; this single file by itself doesn't compress well, but it appears to be the root cause of `mkdwarfs` using an excessive amount of memory?

**Comment by mhx** [comment]
> What I did find instead was that using the -l0 option and then recompressing the image works in these cases without issue. Creating the initial image with -S24 results in very well recompressed files in these instances, the 3.1 GB file compressing down to 2.3 GB whereas the default block size for -l0 resulted in a 2.6 GB file (which is approx what mksquashfs -comp zstd -b 1M -Xcompression-level 22 also gave me).

This mostly makes sense. You can't change the block size (`-S`) during recompression, so that explains why it compresses much better with `-S24`. Nothing is stopping you from using `-l0 -S24` initially if that helps. At `-l0`, the block size is 1MiB, just as with `mksquashfs -b 1M`, so again, no real surprise.

**Comment by Phantop** [comment]
> @Phantop, just to make sure I understand this correctly: you're trying to build DwarFS image from a tree where single files have a size of 3.1GB; this single file by itself doesn't compress well, but it appears to be the root cause of `mkdwarfs` using an excessive amount of memory?

Yeah. The issue boils down to memory usage increasing to ludicrous levels with larger single file sizes, regardless of the size of the entire folder being compressed. I've had it happen in an 11GB folder with the largest file being ~650MB (which I managed to compress using lower memory and job options) and in a 7GB folder with a 1.1GB file (which required me to recompress). It's entirely due to single files.

> This mostly makes sense. You can't change the block size (-S) during recompression, so that explains why it compresses much better with -S24. Nothing is stopping you from using -l0 -S24 initially if that helps. At -l0, the block size is 1MiB, just as with mksquashfs -b 1M, so again, no real surprise.

Yeah I'm aware of all this. just trying to make it clear that I have been able to compress images with these files, just after some workarounds. Really glad you're working on this!

**Comment by mhx** [comment]
@Phantop, I know exactly what the problem is. I just tried compressing my directory of virtualbox images and RSS memory consumption went through the roof (I stopped the process when RSS hit 50 GiB).

It's definitely fixable, but it's not a trivial fix, so it might take a bit of time. In the meantime, I've got a workaround: pass `--blockhash-window-sizes=-`. This may impact compression ratio a bit, but likely not as much as changing the block size. It should definitely fix your problem, though.
